ooniprobe/Model/NetworkMeasurement.mm

```diff
@@ -464,3 +464,221 @@ -(void)on_entry:(const char*)str{
 }
 
 @end
+
+@implementation Whatsapp : NetworkMeasurement
+
+-(id) init {
+    self = [super init];
+    if (self) {
+        self.name = @"whatsapp";
+    }
+    return self;
+}
+
+-(void)run{
+    self.backgroundTask = [[UIApplication sharedApplication] beginBackgroundTaskWithExpirationHandler:^{
+        [[UIApplication sharedApplication] endBackgroundTask:self.backgroundTask];
+        self.backgroundTask = UIBackgroundTaskInvalid;
+    }];
+    [self run_test];
+}
+
+-(void) run_test {
+    mk::nettests::WhatsappTest test;
+    [super init_common:test];
+    test.on_entry([self](std::string s) {
+        [self on_entry:s.c_str()];
+    });
+    test.start([self]() {
+        [self testEnded];
+    });
+}
+
+/*
+ whatsapp: red if "whatsapp_endpoints_status" or "whatsapp_web_status" or "registration_server" are "blocked"
+ docs: https://github.com/TheTorProject/ooni-spec/blob/master/test-specs/ts-018-whatsapp.md#semantics
+ */
+
+-(void)on_entry:(const char*)str{
+    if (str != nil) {
+        if (!self.entry){
+            [TestStorage set_entry:self.test_id];
+            self.entry = TRUE;
+        }
+        NSError *error;
+        NSData *data = [[NSString stringWithUTF8String:str] dataUsingEncoding:NSUTF8StringEncoding];
+        NSDictionary *json = [NSJSONSerialization JSONObjectWithData:data options:0 error:&error];
+        int blocking = ANOMALY_GREEN;
+        
+        //whatsapp_endpoints_status and whatsapp_web_status and registration_server_status must be both false if null (set test anomaly), if at least one is true (set test blocked)
+        NSArray *keys = [[NSArray alloc] initWithObjects:@"whatsapp_endpoints_status", @"whatsapp_web_status", @"registration_server_status", nil];
+        for (NSString *key in keys) {
+            if ([json objectForKey:key]){
+                if ([json objectForKey:key] == [NSNull null]) {
+                    if (blocking < ANOMALY_ORANGE)
+                        blocking = ANOMALY_ORANGE;
+                }
+                else if ([[json objectForKey:key] isEqualToString:@"blocked"]) {
+                    blocking = ANOMALY_RED;
+                }
+            }
+        }
+        
+        //TODO don't know if needed, maybe remove
```

Comment:
Also here, I would ignore "whatsapp_web_failure" and just look at "whatsapp_web_status".